Initial translation of renderToPipeableStream page (#488)
reactjs/id.react.dev: src/content/reference/react-dom/server/renderToPipeableStream.md

<Intro>

> `renderToPipeableStream` me-*render* pohon React menjadi [Node.js Stream](https://nodejs.org/api/stream.html) yang *pipeable*.

Reviewer: Let's introduce it once along with the original word.

```suggestion
`renderToPipeableStream` me-*render* pohon (*tree*) React menjadi [Node.js Stream](https://nodejs.org/api/stream.html) yang *pipeable*.
```

Reviewer 2: Hmmm, how about keep it as it is for now?

Reviewer: - `*pipeable*` to keep
- `pohon` still need to be introduced with `*tree*`

<Note>

> API ini spesifik untuk Node.js. *Environment* dengan [Web Streams,](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) seperti Deno dan lingkungan waktu larian *edge* modern, harus menggunakan [`renderToReadableStream`](/reference/react-dom/server/renderToReadableStream) sebagai gantinya.

Reviewer 2: ```suggestion
API ini spesifik untuk Node.js. *Environment* dengan [Web Streams,](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) seperti Deno dan *edge runtime* modern, harus menggunakan [`renderToReadableStream`](/reference/react-dom/server/renderToReadableStream) sebagai gantinya.
```

  * `onShellReady` **(opsional)**: *Callback* yang diaktifkan tepat setelah [*shell* awal](#specifying-what-goes-into-the-shell) di-*render*. Anda dapat [mengatur kode status](#setting-the-status-code) dan memanggil `pipe` di sini untuk memulai *streaming.* React akan [mengalirkan konten tambahan](#streaming-more-content-as-it-loads) setelah shell bersama tag `<script>` *inline* yang menggantikan *fallback* pemuatan HTML dengan konten.
  * `onShellError` **(opsional)**: *Callback* yang diaktifkan jika ada *error* saat merender *shell* awal.  Ini menerima *error* sebagai argumen. Belum ada bita yang di-emit dari *stream.* `onShellReady` maupun `onAllReady` tidak akan dipanggil. Dengan demikian, Anda dapat [menampilkan *shell HTML fallback.*](#recovering-from-errors-inside-the-shell)
  * `progressiveChunkSize` **(opsional)**: Jumlah bita dalam sebuah *chunk.* [Baca lebih lanjut tentang *default heuristic.*](https://github.com/facebook/react/blob/14c2be8dac2d5482fda8a0906a31d239df8551fc/packages/react-server/src/ReactFizzServer.js#L210-L225)

> #### Hasil yang dikembalikan {/*returns*/}

Reviewer 2: ```suggestion
#### Kembalian {/*returns*/}
```

}
```

Jika *error* terjadi pada komponen `Posts` atau suatu tempat di dalamnya, React akan [mencoba memulihkannya:](/reference/react/Suspense#providing-a-fallback-for-server-errors-and-server-only-content)

1. Mengirimkan fallback pemuatan dari `<Suspense>` boundary terdekat (`PostsGlimmer`) ke HTML halaman.
2. "Menyerah" untuk mencoba merender konten `Posts` di server lagi.
3. Saat kode JavaScript dimuat di klien, React akan *mencoba lagi* me-*render* `Posts` dari sisi klien.

Jika percobaan me-*render* ulang `Posts` dari klien *juga* gagal, React akan melempar kesalahan di klien. Seperti halnya semua error yang terjadi selama rendering, [*error boundary* induk terdekat](/reference/react/Component#static-getderivedstatefromerror) menentukan bagaimana *error* ditampilkan kepada pengguna. Dalam praktiknya, ini berarti bahwa pengguna akan melihat indikator pemuatan hingga dipastikan bahwa *error* tidak dapat dipulihkan.

> Jika percobaan me-*render* ulang `Posts` dari klien berhasil, fallback pemuatan dari server akan diganti dengan *output rendering* klien. Pengguna tidak akan tahu bahwa ada kesalahan server. Namun, callback `onError` server dan callback [`onRecoverableError`](/reference/react-dom/client/hydrateRoot#hydrateroot) klien akan aktif sehingga Anda bisa mendapatkan pemberitahuan tentang error tersebut.

Reviewer: ```suggestion
Jika percobaan me-*render* ulang `Posts` dari klien berhasil, *fallback* pemuatan dari *server* akan diganti dengan *output rendering* klien. Pengguna tidak akan tahu bahwa ada kesalahan *server*. Namun, *callback* `onError` *server* dan *callback* [`onRecoverableError`](/reference/react-dom/client/hydrateRoot#hydrateroot) klien akan aktif sehingga Anda bisa mendapatkan pemberitahuan tentang *error* tersebut.
```

---

### Mengatur kode status {/*setting-the-status-code*/}

> Streaming menimbulkan *tradeoff*. Anda ingin memulai streaming halaman sedini mungkin agar pengguna dapat melihat konten lebih cepat. Namun, begitu Anda memulai streaming, Anda tidak dapat lagi menyetel kode status respons.

Reviewer: ```suggestion
Streaming menimbulkan *tradeoff*. Anda ingin memulai streaming halaman sedini mungkin agar pengguna dapat melihat konten lebih cepat. Namun, begitu Anda memulai *streaming*, Anda tidak dapat lagi menyetel kode status respons.
```

> Dengan [membagi aplikasi Anda](#specifying-what-goes-into-the-shell) ke dalam shell (di atas semua batas `<Suspense>`) dan konten lainnya, Anda telah menyelesaikan sebagian dari masalah ini . Jika *shell* error, Anda akan mendapatkan callback `onShellError` yang memungkinkan Anda menyetel kode status error. Jika tidak, Anda tahu bahwa aplikasi dapat pulih pada klien, sehingga Anda dapat mengirimkan "OK".

Reviewer: ```suggestion
Dengan [membagi aplikasi Anda](#specifying-what-goes-into-the-shell) ke dalam *shell* (di atas semua batas `<Suspense>`) dan konten lainnya, Anda telah menyelesaikan sebagian dari masalah ini. Jika *shell* terjadi *error*, Anda akan mendapatkan *callback* `onShellError` yang memungkinkan Anda menyetel kode status *error*. Jika tidak, Anda tahu bahwa aplikasi dapat pulih pada klien, sehingga Anda dapat mengirimkan "OK".
```

});
```

> Jika sebuah komponen *di luar* shell (yaitu di dalam `<Suspense>` boundary) melontarkan *error,* React tidak akan berhenti merender. Ini berarti callback `onError` akan diaktifkan, tetapi Anda masih akan mendapatkan `onShellReady` alih-alih `onShellError`. Ini karena React akan mencoba memulihkan dari *error* itu dari sisi klien, [seperti yang dijelaskan di atas.](#recovering-from-errors-outside-the-shell)

Reviewer: ```suggestion
Jika sebuah komponen *di luar* shell (yaitu di dalam `<Suspense>` boundary) melontarkan *error,* React tidak akan berhenti me-*render*. Ini berarti *callback* `onError` akan diaktifkan, tetapi Anda masih akan mendapatkan `onShellReady` alih-alih `onShellError`. Ini karena React akan mencoba memulihkan dari *error* itu dari sisi klien, [seperti yang dijelaskan di atas.](#recovering-from-errors-outside-the-shell)
```

});
```

> Ini hanya akan menangkap *error* di luar *shell* yang terjadi saat membuat konten *shell* awal, jadi tidak lengkap. Jika mengetahui apakah terjadi *error* untuk beberapa konten sangat penting, Anda dapat memindahkannya ke dalam shell.

Reviewer: ```suggestion
Ini hanya akan menangkap *error* di luar *shell* yang terjadi saat membuat konten *shell* awal, jadi tidak lengkap. Jika mengetahui apakah terjadi *error* untuk beberapa konten sangat penting, Anda dapat memindahkannya ke dalam *shell*.
```

---

### Menangani error yang berbeda dengan cara yang berbeda {/*handling-different-errors-in-different-ways*/}

> Anda dapat [membuat *subclass* `Error` sendiri](https://javascript.info/custom-errors) dan menggunakan operator [`instanceof`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) untuk memeriksa kesalahan mana yang dilontarkan. Misalnya, Anda dapat menentukan `NotFoundError` khusus dan melontarkannya dari komponen Anda. Kemudian callback `onError`, `onShellReady`, dan `onShellError` dapat melakukan sesuatu yang berbeda bergantung pada jenis kesalahan:

Reviewer: ```suggestion
Anda dapat [membuat *subclass* `Error` sendiri](https://javascript.info/custom-errors) dan menggunakan operator [`instanceof`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) untuk memeriksa kesalahan mana yang dilontarkan. Misalnya, Anda dapat menentukan `NotFoundError` khusus dan melontarkannya dari komponen Anda. Kemudian *callback* `onError`, `onShellReady`, dan `onShellError` dapat melakukan sesuatu yang berbeda bergantung pada jenis kesalahan:
```

});
```

> Perlu diingat bahwa setelah Anda meng-*emit* shell dan memulai streaming, Anda tidak dapat mengubah kode status.

Reviewer: ```suggestion
Perlu diingat bahwa setelah Anda meng-*emit* shell dan memulai *streaming*, Anda tidak dapat mengubah kode status.
```

---

### Menunggu semua konten dimuat untuk crawler dan static generation {/*waiting-for-all-content-to-load-for-crawlers-and-static-generation*/}

> Streaming menawarkan pengalaman pengguna yang lebih baik karena pengguna dapat melihat konten saat tersedia.

Reviewer: ```suggestion
*Streaming* menawarkan pengalaman pengguna yang lebih baik karena pengguna dapat melihat konten saat tersedia.
```

> Namun, saat *crawler* mengunjungi halaman Anda, atau jika Anda membuat halaman pada saat proses *building,* Anda mungkin ingin membiarkan semua konten dimuat terlebih dahulu, lalu menghasilkan *output* final HTML alih-alih menampilkannya secara bertahap.

Reviewer: ```suggestion
Namun, saat *crawler* mengunjungi halaman Anda, atau jika Anda membuat halaman pada saat waktu pembangunan, Anda mungkin ingin membiarkan semua konten dimuat terlebih dahulu, lalu menghasilkan keluaran akhir HTML alih-alih menampilkannya secara bertahap.
```

});
```

> Pengunjung reguler akan mendapatkan *stream* konten yang dimuat secara progresif. *Crawler* akan menerima hasil akhir HTML setelah semua data dimuat. Namun, ini juga berarti *crawler* harus menunggu *semua* data, beberapa di antaranya mungkin lambat dimuat atau error. Bergantung pada aplikasi Anda, Anda juga dapat memilih untuk mengirim shell ke crawler.

Reviewer: - output = keluaran
- result = hasil

```suggestion
Pengunjung reguler akan mendapatkan *stream* konten yang dimuat secara progresif. *Crawler* akan menerima keluargan akhir HTML setelah semua data dimuat. Namun, ini juga berarti *crawler* harus menunggu *semua* data, beberapa di antaranya mungkin lambat dimuat atau error. Bergantung pada aplikasi Anda, Anda juga dapat memilih untuk mengirim *shell* ke *crawler*.
```

}, 10000);
```

> React akan menghapus *fallback* pemuatan yang tersisa sebagai HTML, dan akan mencoba merender sisanya dari sisi klien.

Reviewer: ```suggestion
React akan menghapus *fallback* pemuatan yang tersisa sebagai HTML, dan akan mencoba me-*render* sisanya dari sisi klien.
```
